sympy/tensor/array/array_comprehension.py

```diff
@@ -87,6 +111,8 @@ def _check_bounds_validity(cls, expr, bounds):
                 raise TypeError('Bounds should be an Expression(combination of Integer and Symbol)')
             if isinstance(inf, Integer) and isinstance(sup, Integer) and (inf > sup) == True:
                 raise ValueError('Lower bound should be inferior to upper bound')
+            if var in inf.free_symbols or var in sup.free_symbols:
+                raise ValueError('Variable should not be part of its bounds')
```

Comment:
I think that's quite intuitive to have variable symbols to be replaced rather than constants. Moreover, instead of allowing only `Symbol`, expressions containing `Symbol` should be allowed. Like `(n + 1, 4, 8)`.

Reply:
The variable needs to be a symbol, of course. `inf` and `sup` can be any expression.

> Moreover, instead of allowing only Symbol, expressions containing Symbol should be allowed. Like (n + 1, 4, 8).

No, use `(n, 4+1, 8+1)`. First variable needs to be a symbol.

Have you ever trying a list comprehension like this:
```python
[n for n+1 in range(4, 8)]
```
I don't think this should be allowed.